File: Introductory Problems/16_GridPaths.py
# Working but too Slow
if __name__ == "__main__":
    sol = 0
    ip = str(input())
    visited = [[False for i in range(7)] for j in range(7)]

    def pathFind(i, j, path):
        if (path > 48 or (i == 6 and j == 0)):
            if path == 48:
                global sol
                sol += 1
            return
        global visited
        visited[i][j] = True
        u = (i-1 >= 0) and (not visited[i-1][j])
        d = (i+1 <= 6) and (not visited[i+1][j])
        l = (j-1 >= 0) and (not visited[i][j-1])
        r = (j+1 <= 6) and (not visited[i][j+1])
        if ((not l) and (not r) and u and d) or ((not u) and (not d) and l and r):
            return
        global ip
        if ip[path] != '?':
            if ip[path] == 'U':
                d = l = r = False
            elif ip[path] == 'D':
                u = l = r = False
            elif ip[path] == 'L':
                u = d = r = False
            else:
                u = d = l = False
        if u:
            pathFind(i-1, j, path+1)
            visited[i-1][j] = False
        if d:
            pathFind(i+1, j, path+1)
            visited[i+1][j] = False
        if l:
            pathFind(i, j-1, path+1)
            visited[i][j-1] = False
        if r:
            pathFind(i, j+1, path+1)
            visited[i][j+1] = False

    pathFind(0, 0, 0)
    print(sol)

File: Introductory Problems/16_GridPaths.cpp
#include <bits/stdc++.h>

using namespace std;

bool visited[7][7];
string ip;
int sol = 0;

void pathFind(int row, int col, int len_path = 0)
{
  if (row == 6 && col == 0)
  {
    if (len_path == 48)
    {
      sol++;
    }
    return;
  }
  visited[row][col] = true;
  if (ip[len_path] == '?' || ip[len_path] == 'L')
  {
    if (col && !visited[row][col - 1])
    {
      if (!(!(col - 2 >= 0 && !visited[row][col - 2]) && row + 1 < 7 && !visited[row + 1][col - 1] && row - 1 >= 0 && !visited[row - 1][col - 1]))
      {
        pathFind(row, col - 1, len_path + 1);
      }
    }
  }
  if (ip[len_path] == '?' || ip[len_path] == 'R')
  {
    if (col < 6 && !visited[row][col + 1])
    {
      if (!(!(col + 2 < 7 && !visited[row][col + 2]) && row + 1 < 7 && !visited[row + 1][col + 1] && row - 1 >= 0 && !visited[row - 1][col + 1]))
      {
        pathFind(row, col + 1, len_path + 1);
      }
    }
  }
  if (ip[len_path] == '?' || ip[len_path] == 'U')
  {
    if (row && !visited[row - 1][col])
    {
      if (!(!(row - 2 >= 0 && !visited[row - 2][col]) && col + 1 < 7 && !visited[row - 1][col + 1] && col - 1 >= 0 && !visited[row - 1][col - 1]))
      {
        pathFind(row - 1, col, len_path + 1);
      }
    }
  }
  if (ip[len_path] == '?' || ip[len_path] == 'D')
  {
    if (row < 6 && !visited[row + 1][col])
    {
      if (!(!(row + 2 < 7 && !visited[row + 2][col]) && col + 1 < 7 && !visited[row + 1][col + 1] && col - 1 >= 0 && !visited[row + 1][col - 1]))
      {
        pathFind(row + 1, col, len_path + 1);
      }
    }
  }

  visited[row][col] = false;
}

int main()
{
  cin >> ip;
  pathFind(0, 0);
  cout << sol;
  return 0;
}